mlflow/sklearn/__init__.py

```diff
@@ -668,7 +671,25 @@ def fit_mlflow(self, func_name, *args, **kwargs):
             else:
                 try_mlflow_log(mlflow.log_metric, "training_score", training_score)
 
-        try_mlflow_log(log_model, self, artifact_path="model")
+        SAMPLE_ROWS = 5
+        fit_arg_names = _get_arg_names(self.fit)
+        X_var_name, y_var_name = fit_arg_names[:2]
+        X_sample = _get_Xy(args, kwargs, X_var_name, y_var_name)[0][:SAMPLE_ROWS]
+        try:
+            model_output = self.predict(X_sample) if hasattr(self, "predict") else None
```

Comment:
There are a couple of interesting questions that come to mind here:

1. If the model doesn't have a `predict()` method, should we still be recording the input portion of the inference schema? (I.e., if `predict()` isn't defined, then what inference routine / function is compatible with the stated inputs?)

2. Should we also consider attempting to fetch outputs for models that define `transform()` but not `predict()`? I.e. perhaps it makes sense to do the following:

- Check if `predict()` is defined. If so, output the schema based on `predict()`
- Else, check if `transform()` is defined. If so, output the schema based on `transform()`
- Else, do not output a schema (see point no.1)

This has the benefit of providing schema information for a wider breadth of models. However, there are a couple of downsides to consider: A. when serving / scoring sklearn models as pyfunc, we assume that the `predict()` method is defined and attempt to invoke it; if we output schemas for models that have `transform()` defined but not `predict()`, that may lead to confusing behavior ("why can't I perform pyfunc inference on my model? you claim it has an inference schema!"). B. Given a schema, users need to reason about which method it refers to (it refers to `predict()` if defined, else to `transform()`). Let me know what your thoughts are here. @tomasatdatabricks it would also be great to get your input here!

3. Can we handle the case where models define `fit_predict()` but do not define `predict()`? Interestingly, that seems to apply to a few estimators, such as: https://scikit-learn.org/stable/modules/generated/sklearn.cluster.AgglomerativeClustering.html#sklearn.cluster.AgglomerativeClustering.fit_predict

Reply:
@dbczumar Those are good questions!  And a cool PR @harupy!

We do not really support models without predict method and we should probably not output PyFunc flavor for those at all until we fix it. If we fix it, I think we should map PyFunc.predict to the relevant method (e.g. transform) on the underlying model. I don't think we want to adapt distinction between predict and transform. 

So my take on this is, ideally we want the following: 
1) find out what should PyFunc.predict call on the model. If we don't find what to map to, do not output pyfunc at all.
2) use the method from step 1 to generate the output schema.

Reply:
Makes perfect sense. Thanks @tomasatdatabricks !

Reply:
makes sense! thanks @tomasatdatabricks